### ISSUE TITLE
[workflow] introduce workflow lifecycle events firehose

### ISSUE THREAD

**Issue by olitomlinson** [dapr/dapr#6072]
/area runtime

## Describe the proposal

@cgillum @nyemade-uversky 

From an operational perspective, it makes sense that that we introduce a firehose of Workflow life-cycle events.

These are incredibly valuable for users who want an evented hook / trigger when a life-cycle event occurs.

A perfect and common example of this is subscribing to a firehose of `failed` events. This allows developers and operations people to keep track of critically important workflows and react accordingly, usually by raising alerts.

[This concept is not anything new, and is fully embraced in Azure Durable Function](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-event-publishing?tabs=csharp-script#configure-azure-event-grid-publishing)

**Comment by olitomlinson** [comment]
Bump 

@cgillum how do we get this on the roadmap?

**Comment by cgillum** [comment]
@nyemade-uversky currently manages the roadmap for workflows, so I think we'll need her agreement to take this on.

I agree this would be a useful feature.

**Comment by olitomlinson** [comment]
It would be really advantageous to have this feature available.

I'm trying to test running many workflows at scale, but there is no easy way for me to see the state of my workflows. For example, if I schedule 10,000 workflows, I want to see 10,000 workflows complete eventually. There's no easy way for me to do this.

Having the lifecycle events would allow me to create a subscriber to persist the complete/failed events and easily aggregate them to check that I've achieved my goal.

Once again, lifecycle events are super important for driving monitoring/alerting in a production scenario.

@nyemade-uversky I really think this needs prioritising into the roadmap, ideally 1.12.

**Comment by yaron2** [comment]
> @nyemade-uversky currently manages the roadmap for workflows, so I think we'll need her agreement to take this on.
> 
> I agree this would be a useful feature.

I don't know what roadmap you are referring to, but what goes in 1.12 or later is decided and prioritized solely by @dapr/maintainers-dapr

**Comment by olitomlinson** [comment]
@yaron2 this is the roadmap that we're referring too :) https://github.com/orgs/dapr/projects/61#card-88408739

Sorry, wasn't trying to circumvent any process, but I thought @nyemade-uversky was in the drivers seat for what happens and when with regards to Workflows :) Happy to be wrong on that of course.

**Comment by olitomlinson** [comment]
Bump

**Comment by olitomlinson** [comment]
Another use-case of this feature would be to support the purging of instance data for dapr workflows.

The lifecycle events would be subscribed by the Workflow app itself, which then calls high-level workflow management APIs to purge individual workflow instances when they reach a certain workflow status.

Some users may want to only purge `completed` instances
Others may want to only purge `failed` instances etc.

This allows a flexible model to achieve each of those goals, with no additional complexity for dapr maintainers.

**Comment by olitomlinson** [comment]
Bump

**Comment by olitomlinson** [comment]
bump

**Comment by olitomlinson** [comment]
Bump

**Comment by olitomlinson** [comment]
Bump

**Comment by olitomlinson** [comment]
Another use-case is for building a secondary index of instance IDs which can be used to drive a polling process which checks for versions of workflows which are no longer running I.e when looking to deprecate a particular version of a workflow type from the code base.

**Comment by cgillum** [comment]
Agreed with the secondary index use case, which is a big gap in the current implementation.

Should we aggregate the list of potential use cases in the OP, for easier referencing?

**Comment by olitomlinson** [comment]
> Should we aggregate the list of potential use cases in the OP, for easier referencing?

Yep good idea. done that!

**Comment by oising** [comment]
Agree strongly that this will elevate dapr workflows greatly.

**Comment by ajstewart** [comment]
A +1 from me for this feature, would be very useful!

**Comment by oising** [comment]
La la la, bump.
